contracts/serverless-v2/Executors.sol

```diff
@@ -0,0 +1,509 @@
+    error ExecutorsAlreadyInitiatedDeregister();
+    error ExecutorsAlreadyInitiatedUnstake();
+    error ExecutorsInvalidAmount();
+    error ExecutorsInvalidExecutor();
+
+    //-------------------------------- internal functions start ----------------------------------//
+
+    function _registerExecutor(
+        bytes memory _attestationSignature,
+        IAttestationVerifier.Attestation memory _attestation,
+        uint256 _jobCapacity,
+        bytes memory _signature,
+        uint256 _stakeAmount
+    ) internal {
+        if(_stakeAmount < MIN_STAKE_AMOUNT)
+            revert ExecutorsLessStakeAmount();
+        
+        address operator = _msgSender();
+        if(executors[operator].enclaveKey != address(0))
+            revert ExecutorsExecutorAlreadyExists();
+
+        // attestation verification
+        _verifyEnclaveKey(_attestationSignature, _attestation);
+
+        address enclaveKey = _pubKeyToAddress(_attestation.enclavePubKey);
+        // signature check
+        _verifySign(operator, enclaveKey, _jobCapacity, _signature);
+
+        _register(operator, enclaveKey, _jobCapacity);
+
+        // add node to the tree
+        _insert_unchecked(operator, uint64(_stakeAmount));
+
+        _addStake(operator, _stakeAmount);
+    }
+
+    function _verifySign(
+        address _operator,
+        address _enclaveKey,
+        uint256 _jobCapacity,
+        bytes memory _signature
+    ) internal pure {
+        bytes32 hashStruct = keccak256(
+            abi.encode(
+                REGISTER_TYPEHASH,
+                _operator,
+                _jobCapacity
+            )
+        );
+        bytes32 digest = keccak256(abi.encodePacked("\x19\x01", DOMAIN_SEPARATOR, hashStruct));
+        address signer = digest.recover(_signature);
+
+        if(signer != _enclaveKey)
+            revert ExecutorsInvalidSigner();
+    }
+
+    function _register(
+        address _operator,
+        address _enclaveKey,
+        uint256 _jobCapacity
+    ) internal {
+        executors[_operator].enclaveKey = _enclaveKey;
+        executors[_operator].jobCapacity = _jobCapacity;
+        executors[_operator].status = true;
+        
+        emit ExecutorRegistered(_operator, _enclaveKey);
+    }
+
+    function _deregisterExecutor() internal {
+        address operator = _msgSender();
+        _isValidExecutor(operator);
+        if(!executors[operator].status)
+            revert ExecutorsAlreadyInitiatedDeregister();
+
+        executors[operator].status = false;
+
+        // remove node from the tree
+        _deleteIfPresent(operator);
+
+        if(executors[operator].activeJobs == 0)
+            _completeDeregister(operator);
+        else
+            emit ExecutorDeregisterInitiated(operator);
+    }
+
+    function _addExecutorStake(
+        uint256 _amount
+    ) internal {
+        if(_amount == 0)
+            revert ExecutorsInvalidAmount();
+
+        address operator = _msgSender();
+        _isValidExecutor(operator);
```

Comment:
@roshanr95  so initially I wanted to add it as a modifier only, but then the modifier would also call the `_msgSender()` in the external function apart from setting `address operator = _msgSender();` in the code. Since there were 2 calls happening to fetch _msgSender(),so decided to make it an internal function instead of modifier

Reply:
that's fine

Reply:
changes pushed